docs/dag-sharding-rfc.md

```diff
@@ -0,0 +1,176 @@
+
+The first challenge will require rethinking the go-ipfs dag building functionality and on the bright side may contribute to building a versatile foundation for the potential dagger/dex project.  The second challenge is best addressed by evaluating the performance characteristics of both data transfer methods.  Perhaps the ongoing ipfs metrics work will already answer many questions about the first method.  If the second transfer method (libp2p rpc) is a serious contender we should plan on evaluating its performance as part of implementation.  (TODO -- create an issue defining some metrics that we will want to evaluate) (TODO -- see if any existing ipfs metrics are relavent to first method, e.g. measuring remote pin time of dags with many large leaves)
+
+### Collaborative importing
+We could potentially use alternatives to importing the stream on one master node that then sends out shards to the assigned pinning nodes.  One other possibility is "collaborative importing", where the cluster node doing the "add" operation streams the raw data out to the pinning nodes that then do importing and pinning in place.
+
+One potential downside is that importing the ipfs files dag becomes more complex, as there will in general be complex relationships between the dags of different shards. Note this should technically be possible, especially under the assumption that the entire ipfs file dag fits in all ipfs repos.  After each node runs the import they could combine together their ipfs dags (along with their data's relative order in the original stream) into a clever combining function that could construct the correct final graph on a single node.   This would also likely be more difficult to coordinate among nodes, and may be harder to get right in the face of network failures that require one node to restart downloading.  One final downside is that, depending on the splitter used, it may be difficult to efficiently stream data to peers that fits along the original chunk boundaries.
+
+One upside is that the interaction between the ipfs files dag and cluster-dag importing process becomes much more simple.  The original importer could run to completion uninterrupted, and then the node could build the cluster-dag and unpin the leaves from the ipfs files dag.  Althought this wouldn't directly address the issues brought up above it is possible collaborative importing would result in a simpler download process as now downloads and imports are not interleaved.  It also seems likely that, if done right, collaborative importing could be faster overall, as importing would be parallelized across nodes.
+
+
+## Arbitrary dags
+
+TODO -- I have some thoughts on these (many voiced in my looong comment in response to @hsanjuan's proposal), I need to take some time to organize and record them here
+
+### Dags too big to fit in ipfs repo
+
+### Dags with data not in the leaves
```

Comment:
same as above.

The problem here is that it will make fetching slower because the ipfs daemon will need to fetch not only the data leaves but the intermediary nodes when resolving a path.